# Psychoinformatics - Week 9 (Exercises)
by 楊宗樺 (b05705050@ntu.edu.tw)

In [52]:
import numpy as np
import math
import random
from scipy import stats
from sklearn import *
from sklearn import model_selection
from matplotlib.pyplot import *
%matplotlib inline

## 1 檢查 machine learning pipeline (6 points)

### 1.1 請打亂原本的Y觀察正確率是否和chance level (0.33)有差異? 若有, why? (3 points)

#  1.1-1若在clf.fit(X,Y) 前打亂Y，則

In [53]:
# 本題在研究打亂X和打亂Y有差別嗎?
Accuracy1=[]
for i in range(100):
    iris = datasets.load_iris()
    X=iris.data
    Y=iris.target
    random.shuffle(Y)
    clf=neighbors.KNeighborsClassifier(1)
    clf.fit(X,Y)
    accuracy=np.mean(clf.predict(X)==Y)
    Accuracy1.append(accuracy)

In [54]:
print(max(Accuracy1))
print(min(Accuracy1))

1.0
0.9933333333333333


#### 可看出測試100次的準確率最高為1最低為0.993333，遠高於0.33
#### 原因為predict的data與train的data是相同的，自然地準確率會很高

# 1.1-2若在clf.fit(X,Y) 之後打亂Y，則

In [55]:
# 本題在研究打亂X和打亂Y有差別嗎?
Accuracy2=[]
chance=[0.33]*100
for i in range(100):
    iris = datasets.load_iris()
    X=iris.data
    Y=iris.target
    clf=neighbors.KNeighborsClassifier(1)
    clf.fit(X,Y)
    random.shuffle(Y)
    accuracy=np.mean(clf.predict(X)==Y)
    Accuracy2.append(accuracy)

#### 使用雙尾檢定(alpha=0.01)
#### H0: mean(Accuracy2)=0.33
#### H1: mean(Accuracy2)!=0.33

In [56]:
df=200-2
t = (np.mean(Accuracy2) - np.mean(chance)) / (np.std(Accuracy2) / math.sqrt(100)+(np.std(chance) / math.sqrt(100)))
print(t)
p=stats.t.sf(np.abs(t), df)*2
print("t = " + str(t))
print("p = " + str(p))

-1.3057338354316104
t = -1.3057338354316104
p = 0.19315814340645768


#### p-value=0.19315814340645768>0.01
#### =>It doesn't reject the null hypothesis
#### =>我們有99％的信心水準來說，打亂原本的Y觀察正確率和chance level (0.33)無差異

### 1.2 請用母數或無母數統計檢定以下accuracies中的結果是否和chance level (0.5)有差異? 若有, why? (3 points)

In [47]:
import numpy as np
import math
import random
from scipy import stats
from sklearn import *
from sklearn import model_selection
from matplotlib.pyplot import *
%matplotlib inline

In [48]:
Y=np.remainder(range(200),2) 
print(Y) #Y的0和1個數一樣多

[0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0
 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1
 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0
 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1
 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0
 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1]


In [49]:
# 跑一百次測試:
clf=svm.SVC()
accuracies=[]
for i in range(100): 
 X=np.random.rand(200,2) # X取亂數
 kf=model_selection.KFold(len(Y),shuffle=True) # Leave-one-out cross-validation
 sc=model_selection.cross_val_score(clf,X,Y,cv=kf)
 accuracies.append(sc.mean())

In [50]:
# Please do your statistical tests here:
print(accuracies)
print(np.mean(accuracies))
print(np.std(accuracies))

[0.505, 0.495, 0.54, 0.45, 0.385, 0.54, 0.415, 0.52, 0.46, 0.525, 0.495, 0.455, 0.48, 0.57, 0.53, 0.565, 0.505, 0.545, 0.505, 0.485, 0.435, 0.555, 0.5, 0.49, 0.445, 0.39, 0.495, 0.535, 0.515, 0.47, 0.48, 0.47, 0.545, 0.51, 0.395, 0.485, 0.535, 0.38, 0.56, 0.53, 0.46, 0.565, 0.475, 0.52, 0.56, 0.55, 0.355, 0.315, 0.53, 0.55, 0.375, 0.435, 0.555, 0.48, 0.53, 0.545, 0.445, 0.465, 0.545, 0.53, 0.535, 0.39, 0.4, 0.465, 0.535, 0.525, 0.52, 0.395, 0.45, 0.475, 0.53, 0.555, 0.36, 0.525, 0.535, 0.445, 0.515, 0.57, 0.495, 0.545, 0.535, 0.525, 0.46, 0.44, 0.445, 0.405, 0.515, 0.49, 0.51, 0.42, 0.52, 0.56, 0.42, 0.495, 0.5, 0.505, 0.43, 0.45, 0.555, 0.52]
0.48940000000000006
0.05663161661121816


#### 使用雙尾檢定(alpha=0.01)
#### H0: mean(accuracies)=0.5
#### H1: mean(accuracies)!=0.5

In [51]:
c_level=[0.5]*100
df=200-2
t = (np.mean(accuracies) - np.mean(c_level)) / (np.std(accuracies) / math.sqrt(100)+(np.std(c_level) / math.sqrt(100)))
print(t)
p=stats.t.sf(np.abs(t), df)*2
print("t = " + str(t))
print("p = " + str(p))

-1.8717459670575938
t = -1.8717459670575938
p = 0.06271641135091942


#### p-value=0.06271641135091942>0.01
#### =>It doesn't reject the null hypothesis
#### =>我們有99％的信心水準來說，accuracies中的結果和chance level (0.5)無差異

### Please submit this file to http://hpc.psy.ntu.edu.tw/info before noon next Thursday.